src/bson_mapper/bson_archiver.hpp

```diff
@@ -0,0 +1,1097 @@
+    }
+
+    /**
+     * Designates the most recently added node as finished, which will end
+     * any objects or arrays we happen to be in.
+     */
+    void finishNode() {
+        // Iff we ended up serializing an empty object or array, writeName
+        // will never have been called - therefore the node type at the top
+        // of the stack will be StartArray or StartObject - so start and
+        // then immediately end the documents/array.
+
+        // We'll also end any documents/arrays we happen to be in.
+
+        if (_nodeTypeStack.empty()) {
+            throw bson_mapper::Exception("Attempting to finish a nonexistent node.");
+        }
+
+        switch (_nodeTypeStack.top()) {
+            case OutputNodeType::StartArray:
+                _bsonBuilder.open_array();
+            // Don't break so that this array can also be closed.
+            case OutputNodeType::InArray:
+                _bsonBuilder.close_array();
+                break;
+            case OutputNodeType::StartObject:
+                if (_nodeTypeStack.size() > 1 || _objAsRootElement) {
+                    _bsonBuilder.open_document();
+                }
+            // Don't break so that this document can also be closed.
+            case OutputNodeType::InObject:
+                if (_nodeTypeStack.size() > 1) {
+                    _bsonBuilder.close_document();
+                } else if (_objAsRootElement) {
+                    _bsonBuilder.close_document();
+                }
+                break;
+        }
+
+        _nodeTypeStack.pop();
+        _curNodeInheritsUnderlyingBSONDataBase.pop();
+
+        // Reset the _objAsRootElement tracker bool if the stack was emptied.
+        if (_nodeTypeStack.empty()) {
+            _objAsRootElement = false;
+        }
+        writeDocIfRoot();
+    }
+
+    /**
+     * Sets the name for the next node or element.
+     *
+     * @param name
+     *    The name for the next node or element.
+     */
+    void setNextName(const char* name) {
+        _nextName = name;
+    }
+
+    /**
+     * Saves a value of type T to the current node.
+     *
+     * @param t
+     *    A value t where T is any type that is accepted by an overload
+     *    of _bsonBuilder.append(), except for bsoncxx view types.
+     */
+    template <class T>
+    typename std::enable_if<!is_bson_view<typename std::decay<T>::type>::value>::type saveValue(
+        T&& t) {
+        _bsonBuilder.append(std::forward<T>(t));
+    }
+
+    /**
+     * Saves a value of type T to the current node, and throws an exception if we're serializing a
+     * bsoncxx view type without it being wrapped in a class that inherits UnderlyingBSONDataBase.
+     *
+     * @param t
+     *    A value t where T is one of the bsoncxx view types.
+     */
+    template <class T>
+    typename std::enable_if<is_bson_view<typename std::decay<T>::type>::value>::type saveValue(
```

Comment:
It has to be there, because it gets rid of any const or reference qualifiers before comparing it to the types in the is_bson_view struct. Without it, the correct overload isn't chosen.